File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode mergeSort(ListNode l1 , ListNode l2){
        ListNode dummy = new ListNode(-1);
        ListNode temp = dummy;

        while(l1 != null && l2 != null){
            if(l1.val <= l2.val){
                temp.next = l1;
                l1 = l1.next;
            } else {
                temp.next = l2;
                l2 = l2.next;
            }

            temp = temp.next;
        }

        temp.next = l1 != null ? l1 : l2;

        return dummy.next;
    }
    public ListNode mergeKLists(ListNode[] lists) {
        int len = lists.length;

        if(len == 0 || (lists[0] == null && len == 1))
            return null;

        ListNode ans = lists[0];

        for(int i = 1 ; i < lists.length; i++){
            ans = mergeSort(ans , lists[i]);
        }

        return ans;
    }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def merge(self , list1: Optional[ListNode] , list2: Optional[ListNode]):
        p1 , p2 = list1 , list2

        ans = ListNode(0)
        temp = ans

        while p1 and p2 :
            if p1.val < p2.val :
                temp.next = p1
                temp = temp.next
                p1 = p1.next

            else :
                temp.next = p2
                temp = temp.next
                p2 = p2.next

        temp.next = p1 if p1 else p2

        return ans.next

    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        n = len(lists)

        if n == 0 or (n == 1 and lists[0] == None):
            return None

        ans = lists[0]

        for x in range(1 , n):
            ans = self.merge(ans , lists[x])

        return ans